### ISSUE TITLE
Stack overflow if dev dependency has links

### ISSUE THREAD

**Issue by dtolnay** [rust-lang/cargo#8966]
Minimal repro:

```toml
# Cargo.toml

[package]
name = "repro"
version = "0.0.0"
links = "repro"

[dev-dependencies]
dep = { path = "dep" }
```

```rust
// src/lib.rs (empty)
```

```rust
// build.rs

fn main() {}
```

```toml
# dep/Cargo.toml

[package]
name = "dep"
version = "0.0.0"
links = "dep"

[dependencies]
repro = { path = ".." }
```

```rust
// dep/src/lib.rs (empty)
```

```rust
// dep/build.rs

fn main() {}
```

---

```console
$ cargo check --tests

thread 'main' has overflowed its stack
fatal runtime error: stack overflow
Aborted (core dumped)
```

This crashes quite early on in Cargo. A `target` directory isn't even created.

This is a regression introduced in whatever Cargo shipped with Rust 1.28.0. Earlier toolchains build this successfully.

```console
$ cargo +1.27.0 check --tests
   Compiling repro v0.0.0 (file:///git/repro)
   Compiling dep v0.0.0 (file:///git/repro/dep)
    Finished dev [unoptimized + debuginfo] target(s) in 0.19s
```

**Comment by Eh2406** [comment]
Fascinating! I am very much looking forward to digging in on this. I'm wondering how this has gone unnoticed for so long.

**Comment by alexcrichton** [comment]
Wow we are really bad about stack overflowing with build scripts. We seem to never be able to get this quite right... With a blast from the past I've found that the stack overflow is in [this function](https://github.com/rust-lang/cargo/blob/d274fcf862b89264fa2c6b917b15230705257317/src/cargo/ops/cargo_compile.rs#L1327), but I don't believe that function is at fault. The problem is that we have a cycle in the dependency graph which says that the execution of repro's build script depends on the execution of dep's build script, and vice versa.

The correct dependency graph would have dep's build script depend on repro's, but that's it. Build scripts shouldn't depend on the build scripts off dev-dependencies, only normal dependencies. Turns out the bug is [basically here](https://github.com/rust-lang/cargo/blob/d274fcf862b89264fa2c6b917b15230705257317/src/cargo/core/compiler/unit_dependencies.rs#L668-L673) where we *also* need to filter for dev-dependencies.

That function traces its history back to https://github.com/rust-lang/cargo/pull/5711 (spawned from https://github.com/rust-lang/cargo/issues/5708, more stack overflows). That function's logic didn't start in that commit though and traces back further, but bisection locally doesn't seem to be helping. In any case this has clearly been in Cargo for a long time, and I think it makes sense that this situation doesn't arise that much in practice because `links` libraries are pretty far apart, much less depending on each other.

In any case definitely still something that we should fix!
